File: ABC273/d.py
def main():
    h, w, r0, c0 = map(int, input().split())
    n = int(input())
    walls = [tuple(map(int, input().split())) for _ in range(n)]
    iswall = [[False] * w for _ in range(h)]
    for r, c in walls:
        iswall[r - 1][c - 1] = True
    r, c = r0 - 1, c0 - 1
    q = int(input())
    for _ in range(q):
        d, l = input().split()
        l = int(l)
        dr, dc = 0, 0
        if d == 'U':
            dr = -1
        elif d == 'D':
            dr = 1
        elif d == 'L':
            dc = -1
        elif d == 'R':
            dc = 1
        for _ in range(l):
            r += dr
            c += dc
            if 0 <= r < h and 0 <= c < w and not iswall[r][c]:
                continue
            r -= dr
            c -= dc
            break
        print(r + 1, c + 1)


if __name__ == '__main__':
    main()

File: ABC273/d.rs
fn main() {
    proconio::input! {
        h: usize, w: usize, r: usize, c: usize,
        n: usize,
        walls: [(usize, usize); n],
        q: usize,
        queries: [(char, u64); q],
    }
    let (mut r, mut c) = (r - 1, c - 1);
    let mut iswall = vec![vec![false; w]; h];
    for (i, j) in walls {
        iswall[i - 1][j - 1] = true;
    }
    for (d, l) in queries {
        let (dr, dc) = match d {
            'U' => (-1, 0),
            'D' => (1, 0),
            'L' => (0, -1),
            'R' => (0, 1),
            _ => unreachable!(),
        };
        let mut nr = r as i64;
        let mut nc = c as i64;
        for _ in 0..l {
            nr += dr;
            nc += dc;
            if (nr < 0 || nr >= h as i64)
                || (nc < 0 || nc >= w as i64)
                || iswall[nr as usize][nc as usize]
            {
                nr -= dr;
                nc -= dc;
                break;
            }
        }
        r = nr as usize;
        c = nc as usize;
        println!("{} {}", r + 1, c + 1);
    }
}
